decidim-core/lib/decidim/gamification/badge.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Gamification
+    # This class represents an abstract badge. Instances of this class can define
+    # different badge types with different rules such as gaining new levels, etc.
+    class Badge
+      include Virtus.model
+      include ActiveModel::Validations
+
+      # The name of the badge.
+      attribute :name, String
+
+      # An array of scores needed to reach a new level. For example, the array
+      # [1, 5, 10] represents 1 point to get to Level 1, 5 points to get to level 2,
+      # 10 points to get to level 3.
+      attribute :levels, Array, default: []
+
+      # (Optional) you can set a lambda in order to be able to reset the score of a
+      # badge if the progress gets lost somehow. The lambda receives a user as an
+      # argument.
+      #
+      # It might not be possible sometimes, so it's fine to leave it empty.
+      attribute :reset, Proc
+
+      validates :name, :levels, presence: true
+      validates :levels, empty: false
+
+      validate do
+        errors.add(:levels, "level thresholds should be ordered") if levels.sort != levels
```

Comment:
I18n?

Reply:
These messages aren'tt meant to reach user interfaces - it's an internal consistency validation. I don't think I18n make sense here.